### DISCUSSION TITLE
graphql update page mutation: Cannot read property 'map' of undefined

### DISCUSSION THREAD

**Discussion by triffid** [requarks/wiki#3627] [Error / Bug Report]
**Describe the bug**
Getting " Cannot read property 'map' of undefined" for page update mutations

**To Reproduce**
Feeding

	mutation {
		pages {
			update (id:2, content:"This page edited via API") {
				responseResult {
					errorCode,
					message,
					succeeded,
					slug
				}
			}
		}
	}

into https://my-wiki-host/graphql via web browser gives me

	{
		"data": {
			"pages": {
				"update": {
					"responseResult": {
						"errorCode": 1,
						"message": "Cannot read property 'map' of undefined",
						"succeeded": false,
						"slug": "TypeError"
					}
				}
			}
		}
	}

Which seems like either an internal error, or a message that's just as obtuse as the one that appears when I forget to put the authorization token into http headers.

**Expected behavior**
Either page text gets changed, or receive a _meaningful_ error message that suggests the proper way to do it.

**Host Info (please complete the following information):**
 - OS: Docker Container (Linux)
 - Wiki.js version: 2.5.170
 - Database engine: PostgreSQL 11.8

**Additional context**
I'm having a terrible time trying to find any documentation for this API - seems like there's a machine-readable schema but no tutorials or examples anywhere, so I'm left stumbling about in the dark trying to guess what query format I need to do what I want.

**Comment by NGPixel** [comment]
The API is self-documented using GraphQL. You can go to `/graphql` on your installation to launch the GraphQL Playground. You can build queries from there and see the full documentation on the right.

**Reply by JCOvergaar** [reply]
It's only self-documented if you can actually access it, which you can't if the system is set to production. Would it be so difficult to paste a copy of the schema into the API page?

**Comment by triffid** [comment]
A list of functions and parameters is hardly documentation, and it offers zero clue why I might be getting "Cannot read property 'map' of undefined" back from the server for what seems like a fairly simple query.

**Comment by a-stoyanov** [comment]
I've also ran across this issue and it seems to be a bug. It goes away when you declare the id argument as a **Int!** type variable and pass it to the mutation like so:

```console
mutation Page(
  $id: Int!
  $content: String!
  $description: String!
  $tags: [String]!
  $title: String!
  $isPublished: Boolean!
) {
  pages {
    update(
      id: $id
      content: $content
      description: $description
      tags: $tags
      title: $title
      isPublished: $isPublished
    ) {
      responseResult {
        succeeded
        errorCode
        slug
        message
      }
      page {
        id
        path
        title
      }
    }
  }
}
```

VARS
```console
{
"id": 21, 
"content": "HELLO WORLD!",
"title": "page title",
"description": "page description",
"tags": ["patching", "maintenance"],
"isPublished": true
}
```

**Reply by DavidArno** [reply]
This didn't unfortunately work for me. It was @xrisk's tip about needed to specify tags that fixed this for me.

**Comment by xrisk** [comment]
This is definitely a bug. Specifically, if `tags` is not passed it always throws an error. Presumably, the code tries to map over `tags`, but we haven’t passed it.

**Reply by whenSunSet** [reply]
so Is there a plan to fix this bug

**Comment by FreeWillyB** [comment]
So here it is in February 2022 and this is still happening for me. Not only do I get the above error, but I can't even update the title or anything without including the content. And if I include the content, then I get the error of "Cannot read property 'map' of undefined". But as @xrisk pointed out, adding the "tags" argument clears it up. 

So all this means that in order to change the Title of a post, my code goes from 

```
mutation {
  pages {
    update( 
      id:8, 
      title:"New Title"
    ) {
      responseResult {succeeded,errorCode,slug,message}
    }
  }
}
```

to this...

```
mutation {
  pages {
    update(      
      id:8, 
      title:"New Title", 
      content:"<Insert long, existing content here and hope it's exactly the same!>", 
      tags:""
    ) {
      responseResult {succeeded,errorCode,slug,message}
    }
  }
}
````

**Comment by DavidArno** [comment]
Adding my voice to the requests that this bug be fixed. @NGPixel's "*The API is self-documented using GraphQL*" comment is deeply unhelpful here as the API clearly does not document this bug, nor why it still exists three years later.

To add further to the solution provided by @xrisk and @FreeWillyB, in order to eg only update the title of a page, first you need to query the page to get the current content and tags. Then you must supply that content and tags data when calling the update. If you don't supply them, you get errors (a completely cryptic one in the case of tags) and if you supply them, they must be the required data as those elements of the page are updated via the call.

Is the fix to this bug to simply mark those fields as not null (`!`) in the schema?

**Comment by tombayo** [comment]
Just ran into this issue myself whilst working towards the GraphQL-api. Adding `tags` solved the issue in my case.
While investigating the issue I came across another Issue created in this project related to this, which also had been converted into a Discussion:
* https://github.com/requarks/wiki/discussions/3634

However this (the one I'm commenting in) seemed the most comprehensive and active, and given this is clearly a *bug*, might I suggest this Discussion is converted back to an Issue?

Issues can only be created by Project Maintainers, in accordance to https://github.com/requarks/wiki/issues/6044.
@NGPixel : can you (or a project maintainer) please move forward with this bug?
